nfs/src/nfs_client.cpp

```diff
@@ -346,6 +346,72 @@ void nfs_client::readdirplus(
     tsk->run_readdirplus();
 }
 
+void nfs_client::read(
+    fuse_req_t req,
+    fuse_ino_t ino,
+    size_t size,
+    off_t off,
+    struct fuse_file_info *fi)
+{
+    struct rpc_task *tsk1 = rpc_task_helper->alloc_rpc_task();
+    tsk1->init_readfile(req, ino, size, off, fi);
+  
+    // Get the readahead state for this inode.
+    auto rastate = get_nfs_inode_from_ino(ino)->readahead_state;
+
+    // Update the ra state about the application read request received.
+    rastate->on_application_read(off, size);
+
+    // Issue the readfile task.
+    tsk1->run_readfile();
+
+    /*
+     * Once we are done issuing the application read, issue the reads for
+     * readahead. This is done with the help of async rpc task and the response of
+     * the actual read request can be returned back to the caller without having to
+     * wait for readahead reads.
+     */
+
+    /*
+     * The max bytes that can be read from the server is 3MB currently, hence issue
+     * readaheads with this length to get best performance.
+     * TODO: This can be changed later when the blob server supports bigger size RPCs.
+     */
+    const size_t ra_buffer_length = 3 * 1024 * 1024;
```

Comment:
I'm confused how this is still 3MB. Didn't you say that you have changed it to 1MB to prevent crashes?

Anyways, I've submitted a change to ra_state so that we don't need to pass the ra buffer size to get_next_ra() everytime. We initiailize a default ra size in the ra_state constructor and then get_next_ra() uses that.

And to get the ra size correctly supported by libnfs, I submitted a change, now you cal call nfs_get_readmax() and use that as the ra size

Reply:
I used 1MB for my testing, did not push that change since the throughput on single connection was better with this 3MB. I can update it to 1MB.